src/realm/alloc.hpp

```diff
@@ -35,6 +35,10 @@ class Replication;
 
 using ref_type = size_t;
 
+// Maximum .realm file size and also maximum in-memory transaction snapshot size. On 32-bit architectures we need 
+// to cap it to 2^31 * 0.80 bytes because arithmetic on references can otherwise overflow in various places.
+constexpr size_t maximum_realm_size = static_cast<size_t>((std::numeric_limits<ref_type>::max() >> 1) * 0.8);
```

Comment:
By size multiplies, I meant that there may be cases where instead of small additions to the file size we might get a multiple: for example in the case where the allocator doubles the size of the largest allocated chunk. In these cases, if the new size is larger than the "%100" limit on a 32 bit system (max 32 bit unsigned) then the value will overflow and the check `if (ref + new_size > maximum_realm_size)` will actually succeed even though it shouldn't. The concept of an "%80" check only catches overflows between 80-100 percent of the maximum; on the low end it limits previously functioning apps by artificially capping the size and on the high end it doesn't account for overflows past %100 of the actual limit. Increasing or decreasing the %80 buffer compromises on one side or the other, there is no way to win. Therefore I prefer overflow checks wherever possible.